tsdb/ooo_head.go

```diff
@@ -111,22 +112,27 @@ type OOORangeHead struct {
 	// the timerange of the query and having preexisting pointers to the first
 	// and last timestamp help with that.
 	mint, maxt int64
+
+	isoState *oooIsolationState
 }
 
-func NewOOORangeHead(head *Head, mint, maxt int64) *OOORangeHead {
+func NewOOORangeHead(head *Head, mint, maxt int64, minRef chunks.ChunkDiskMapperRef) *OOORangeHead {
+	isoState := head.oooIso.TrackReadAfter(minRef)
```

Comment:
We start tracking the read in `NewOOORangeHead()` and remove it when closing the chunk reader returned from `Chunks()`. If anything errors between `NewOOORangeHead()` and `Chunks()` then we'd never stop tracking this ref. 

This isn't actually a problem right now - nothing can actually error between `NewOOORangeHead()` and `Chunks()`. However, you have to go through a fair bit of code to verify this and it could be easy for a bug to be introduced so I think we should still protect against that case. One way to do that would be to close the `isoState` if `NewBlockQuerier()` errors.

Reply:
Good idea, I've added this in abb21d6.